File: 122. Best Time to Buy and Sell Stock II-2.cpp
class Solution {
public:
  int maxProfit(vector<int> &prices) {
    int hold = -1e4, rest = 0;
    for (auto x : prices) {
      int prevHold = hold;
      hold = max(hold, rest - x);
      rest = max(rest, prevHold + x);
    }

    return rest;
  }
};

File: 122. Best Time to Buy and Sell Stock II-2.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)

        f = [[0] * 2 for _ in range(n + 1)]
        f[0][1] = -inf
        for i, p in enumerate(prices):
            f[i + 1][0] = max(f[i][0], f[i][1] + p)
            f[i + 1][1] = max(f[i][1], f[i][0] - p)

        return f[n][0]
